runtime/src/commonMain/kotlin/pbandk/internal/json/JsonValueUnmarshaller.kt

```diff
@@ -133,4 +135,31 @@ internal class JsonValueUnmarshaller(private val jsonConfig: JsonConfig) {
     } catch (e: Exception) {
         throw InvalidProtocolBufferException("map field did not contain a valid object", e)
     }
+
+    fun readStruct(value: JsonElement): Struct = try {
+        val fields = readMap(value, Struct.fieldDescriptors[0].type as FieldDescriptor.Type.Map<*, *>)
+        Struct(fields.associate { it.key as String to it.value as Value })
+    }  catch (e: InvalidProtocolBufferException) {
+        throw e
+    } catch (e: Exception) {
+        throw InvalidProtocolBufferException("struct field did not contain a valid object", e)
+    }
+
+    fun readValue(value: JsonElement): Value = when {
+        value.isNull -> Value(Value.Kind.NullValue())
+        try { value.primitive; true } catch (e: JsonException) { false } -> {
+            val literal = value.primitive as JsonLiteral
+            when {
+                literal.isString -> Value(Value.Kind.StringValue(value.content))
+                literal.doubleOrNull != null -> Value(Value.Kind.NumberValue(value.double))
+                literal.booleanOrNull != null -> Value(Value.Kind.BoolValue(value.boolean))
+                else -> Value(Value.Kind.StringValue(value.content))
+            }
+        }
+        try { value.jsonArray; true } catch (e: JsonException) { false } ->
+            Value(Value.Kind.ListValue(ListValue(value.jsonArray.map { readValue(it) })))
+        try { value.jsonObject; true } catch (e: JsonException) { false } ->
+            Value(Value.Kind.StructValue(Struct(value.jsonObject.asSequence().associate { (k, v) -> k to readValue(v) })))
```

Comment:
🔧 Similarly here, use the existing code that knows how to read `Struct` already:

```kt
Value(kind = Value.Kind.StructValue(readValue(value, FieldDescriptor.Type.Message(Struct.Companion))))
```